tests/integration-tests/clusters_factory.py

```diff
@@ -0,0 +1,136 @@
+        """Return the name of the CloudFormation stack associated to the cluster."""
+        return "parallelcluster-" + self.name
+
+    @property
+    def region(self):
+        """Return the aws region the cluster is created in."""
+        return self.config.get("aws", "aws_region_name", fallback="us-east-1")
+
+    @property
+    def master_ip(self):
+        """Return the public ip of the cluster master node."""
+        return self.cfn_outputs["MasterPublicIP"]
+
+    @property
+    def os(self):
+        """Return the os used for the cluster."""
+        cluster_template = self.config.get("global", "cluster_template", fallback="default")
+        return self.config.get("cluster {0}".format(cluster_template), "base_os", fallback="alinux")
+
+    @property
+    def cfn_outputs(self):
+        """
+        Return the CloudFormation stack outputs for the cluster.
+        Outputs are retrieved only once and then cached.
+        """
+        if self.__cfn_outputs:
+            return self.__cfn_outputs
+        self.__cfn_outputs = self.__retrieve_cfn_outputs()
+        return self.__cfn_outputs
+
+    @retry(wait_exponential_multiplier=500, wait_exponential_max=5000, stop_max_attempt_number=5)
+    def __retrieve_cfn_outputs(self):
+        logging.debug("Retrieving stack outputs for stack {}".format(self.cfn_name))
+        try:
+            cfn = boto3.client("cloudformation", region_name=self.region)
+            stack = cfn.describe_stacks(StackName=self.cfn_name).get("Stacks")[0]
+            outputs = {}
+            for output in stack.get("Outputs", []):
+                outputs[output.get("OutputKey")] = output.get("OutputValue")
+            return outputs
+        except Exception as e:
+            logging.warning("Failed retrieving stack outputs for stack {} with exception: {}".format(self.cfn_name, e))
+            raise
+
+
+class ClustersFactory:
+    """Manage creation and destruction of pcluster clusters."""
+
+    def __init__(self, ssh_key, test_name=None):
+        self.__test_name = test_name
+        self.__ssh_key = ssh_key
+        self.__created_clusters = {}
+
+    def create_cluster(self, config, name=None):
+        """
+        Create a cluster with a given config.
+        :param config: config file to use for cluster creation.
+        :param name: name of the cluster. If not specified it defaults to "integ-tests-" + random_alphanumeric()
+        :return:
+        """
+        if not name:
+            name = "integ-tests-" + random_alphanumeric()
+
+        if name in self.__created_clusters:
+            raise ValueError("Cluster {0} already exists".format(name))
+
+        # create the cluster
+        logging.info("Creating cluster {0} with config {1}".format(name, config))
+        self.__created_clusters[name] = Cluster(name, config, self.__ssh_key)
+        result = run_command(["pcluster", "create", "--config", config, name])
+        if "CREATE_COMPLETE" not in result.stdout:
```

Comment:
You could define an enum for the cluster status.

Reply:
good idea. I'll apply this in a later review